From aa44c965fc18355686b7c60d60875d483a102982 Mon Sep 17 00:00:00 2001
From: Christian Bruel <christian.bruel@foss.st.com>
Date: Thu, 27 Mar 2025 13:21:46 +0100
Subject: [PATCH] PCI: stm32: use threaded irq handler for AER

AER irq should call the isr after the processing is done and aer mechanism
reset. without this the irq keeps firing without leaving the isr a chance
to consume and clear it.

Use a thread_irq to make sure that we deassert the aer_err_msi interrupt
logic after the aer status is cleared.

Signed-off-by: Christian Bruel <christian.bruel@foss.st.com>
Change-Id: Ic98c66e4d48d1b436fe310fd889bf56c3e94978a
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/447094
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Fabrice GASNIER <fabrice.gasnier@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Fabrice GASNIER <fabrice.gasnier@foss.st.com>
---
 drivers/pci/controller/dwc/pcie-stm32.c | 15 ++++++++++++---
 1 file changed, 12 insertions(+), 3 deletions(-)

diff --git a/drivers/pci/controller/dwc/pcie-stm32.c b/drivers/pci/controller/dwc/pcie-stm32.c
index 8187d4bce760..96976d8c57c8 100644
--- a/drivers/pci/controller/dwc/pcie-stm32.c
+++ b/drivers/pci/controller/dwc/pcie-stm32.c
@@ -242,6 +242,14 @@ static irqreturn_t stm32_pcie_aer_msi_irq_handler(int irq, void *priv)
 	struct stm32_pcie *stm32_pcie = priv;
 
 	regmap_write(stm32_pcie->regmap, SYSCFG_PCIEAERRCMSICR, 1);
+
+	return IRQ_WAKE_THREAD;
+}
+
+static irqreturn_t stm32_pcie_aer_msi_isr_handler(int irq, void *priv)
+{
+	struct stm32_pcie *stm32_pcie = priv;
+
 	regmap_write(stm32_pcie->regmap, SYSCFG_PCIEAERRCMSICR, 0);
 
 	return IRQ_HANDLED;
@@ -304,9 +312,10 @@ static int stm32_add_pcie_port(struct stm32_pcie *stm32_pcie,
 		}
 
 		if (stm32_pcie->aer_irq) {
-			ret = devm_request_irq(dev, stm32_pcie->aer_irq,
-					       stm32_pcie_aer_msi_irq_handler,
-					       IRQF_SHARED, "stm32-aer-msi", stm32_pcie);
+			ret = devm_request_threaded_irq(dev, stm32_pcie->aer_irq,
+							stm32_pcie_aer_msi_irq_handler,
+							stm32_pcie_aer_msi_isr_handler,
+							IRQF_SHARED, "stm32-aer-msi", stm32_pcie);
 			if (ret < 0) {
 				dev_err(dev, "failed to request AER MSI IRQ %d\n",
 					stm32_pcie->aer_irq);
-- 
2.50.1

